src/Pate/Verification/Widening.hs

```diff
@@ -0,0 +1,571 @@
+               widenLoop sym bak (Gas (i-1)) eqCtx postD' (Just newlocs)
+
+
+widenUsingCounterexample ::
+  ( sym ~ W4.ExprBuilder t st fs
+  , PA.ValidArch arch ) =>
+  sym ->
+  W4.GroundEvalFn t ->
+  SimBundle sym arch ->
+  EquivContext sym arch ->
+  W4.Pred sym ->
+  PEE.EquivalenceDomain sym arch ->
+  AbstractDomain sym arch ->
+  Maybe (WidenLocs sym arch) ->
+  AbstractDomain sym arch ->
+  EquivM sym arch (WidenResult sym arch)
+widenUsingCounterexample sym evalFn bundle eqCtx postCondAsm postCondStatePred preD mlocs postD =
+  tryWidenings
+    [ liftIO (widenRegisters sym evalFn bundle eqCtx postCondAsm postCondStatePred postD)
+    , liftIO (widenStack sym evalFn bundle eqCtx postCondAsm postCondStatePred preD postD)
+    , liftIO (widenHeap sym evalFn bundle eqCtx postCondAsm postCondStatePred preD postD)
+    , do slice <- PP.simBundleToSlice bundle
+         ineqRes <- PP.getInequivalenceResult PEE.InvalidPostState (PS.specBody preD) (PS.specBody postD) slice (SymGroundEvalFn evalFn)
+         let msg = unlines [ "Could not find any values to widen!"
+                           , show (pretty ineqRes)
+                           ]
+         return $ WideningError msg (fromMaybe mempty mlocs) postD
+    ]
+
+-- TODO, lots of code duplication between the stack and heap cases.
+--  should we find some generalization?
+widenHeap ::
+  ( sym ~ W4.ExprBuilder t st fs
+  , PA.ValidArch arch ) =>
+  sym ->
+  W4.GroundEvalFn t ->
+  SimBundle sym arch ->
+  EquivContext sym arch ->
+  W4.Pred sym ->
+  PEE.EquivalenceDomain sym arch ->
+  AbstractDomain sym arch ->
+  AbstractDomain sym arch ->
+  IO (WidenResult sym arch)
+-- TODO? should we be using postCondAsm and postConstStatePred?
+widenHeap sym evalFn bundle eqCtx _postCondAsm _postCondStatePred preD postD =
+  do xs <- findUnequalHeapMemCells sym evalFn bundle eqCtx preD
+     ys <- findUnequalHeapWrites sym evalFn bundle eqCtx
+     let zs = xs++ys
+     if null zs then
+       return NoWideningRequired
+     else
+       do -- TODO, this could maybe be less aggressive
+          newCells <- PMc.predFromList sym [ (c, W4.truePred sym) | c <- zs ]
+          let heapDom = PEM.memDomainPred (PEE.eqDomainGlobalMemory (PS.specBody postD))
+          heapDom' <- PMc.mergeMemCellPred sym heapDom newCells
+          let md' = (PEE.eqDomainGlobalMemory (PS.specBody postD)){ PEM.memDomainPred = heapDom' }
+          let pred' = (PS.specBody postD){ PEE.eqDomainGlobalMemory = md' }
+          let postD' = postD{ PS.specBody = pred' }
+          return (Widen (WidenLocs mempty (Set.fromList zs)) postD')
+
+widenStack ::
+  ( sym ~ W4.ExprBuilder t st fs
+  , PA.ValidArch arch ) =>
+  sym ->
+  W4.GroundEvalFn t ->
+  SimBundle sym arch ->
+  EquivContext sym arch ->
+  W4.Pred sym ->
+  PEE.EquivalenceDomain sym arch ->
+  AbstractDomain sym arch ->
+  AbstractDomain sym arch ->
+  IO (WidenResult sym arch)
+-- TODO? should we be using postCondAsm and postConstStatePred?
```

Comment:
I think you want to use `postCondStatePred` to avoid considering locations as "unequal" if they aren't included in the target post-domain.

Reply:
In fact you may want to pass in the intermediate `MemoryCondition` instead, which captures the result of `resolveCellEquiv` or `resolveRegisterEquiv`, but conditional on being in the target domain..

Reply:
Does that really make sense? This is where we are deciding what things we need to _add_ to the target domain.

Reply:
Aren't you technically removing them from the target domain? (i.e. we are weakening the post-condition to exclude locations which aren't equal). In which case I think what you'll end up with is each widening step including locations to widen that are already excluded by the domain (i.e. have already been excluded in a previous widening iteration).

Reply:
I guess it depends on how you think about it.  In this algorithm, the domains always start equating everything, and you progressively add exceptions where things don't have to be equal, so I think about the process of widening as adding new locations.

 At any rate, I think we're already doing what needs to be done via the `assume` statement here:

https://github.com/GaloisInc/pate/blob/f91ac87946fdd10b984bd48a5a93b2985641f3d3/src/Pate/Verification/Widening.hs#L288

But maybe I'm missing something.  I don't really understand the difference between `eqDomPost` and the other things you mentioned.